### PR TITLE
[ci skip] Enhanced RDoc for Array

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Methods:
- #to_a
- #to_h
- #to_ary
- #reverse!
- #reverse

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
Some may say that copied/not-copied goes without saying.  I agree with Tallyrand:  "If it goes without saying, it goes all the better for saying it."

File: array.c [hunk 1]
```diff
@@ -3126,20 +3128,49 @@ rb_ary_to_a(VALUE ary)
 
 /*
  *  call-seq:
- *     ary.to_h                  -> hash
- *     ary.to_h {|item| block }  -> hash
+ *    array.to_h -> new_hash
+ *    array.to_h {|2_element_array| ... } -> new_hash
```
Comment (drbrain) [inline array.c:3132]: What about `|key, value|` instead of `|2_element_array|`? It would save reading down a bit to learn what the two elements were

File: array.c [hunk 2]
```diff
@@ -3207,13 +3253,14 @@ rb_ary_reverse(VALUE ary)
...
  *  call-seq:
- *     ary.reverse!   -> ary
+ *    array.reverse! -> self
  *
- *  Reverses +self+ in place.
- *
- *     a = [ "a", "b", "c" ]
- *     a.reverse!       #=> ["c", "b", "a"]
- *     a                #=> ["c", "b", "a"]
+ *  Returns +self+ with elements (not copies) in reverse order:
```
Comment (drbrain) [inline array.c:3258]: I like the "in place" from the prior text because I think it helps distinguish #reverse and #reverse!

I think the "Returns +self+" part needs to stay to distinguish #reverse!'s behavior from #uniq! and #compact!.

I'm not sure how best to phrase the combination though, thoughts?

File: array.c (hunk 1 above)
Comment (jeremyevans) [reply array.c:3132]: Aren't both `2 element_array` and `key, value` as block arguments wrong?  The block is yielded each element in the array, and should return a 2 element array.  The previous `item` was more accurate.

Comment (drbrain) [reply array.c:3132]: I think you're right and that I confused myself with the way block parameters destructure the input item:

```
$ ruby -e 'p [["a", 1], ["b", 2]].to_h { |k, v| [k.upcase, v] }'
{"A"=>1, "B"=>2}
$ ruby -e 'p %w[a b].to_h { |e| [e.upcase, e] }'
{"A"=>"a", "B"=>"b"}
```

I agree and think this should be something like `array.to_h {|item| ... } -> new_hash`

Perhaps an example like the last one should be added where an array of single elements is converted to a Hash?

File: array.c [hunk 3]
```diff
@@ -3168,9 +3199,24 @@ rb_ary_to_h(VALUE ary)
...
+ *    a1.equal?(a) # => true # Returned self
+ *
+ *  Otherwise, when <tt>self.kind_of?(Array)</tt>,
+ *  returns a new \Array containing the actual elements (not copies) of +self+:
+ *    class MyArray < Array; end
+ *    a = MyArray.new(['foo', 'bar', 'two'])
+ *    a.instance_of?(Array) # => false
+ *    a.kind_of?(Array) # => true
+ *    a1 = a.to_ary
+ *    a1.class # => Array # Not MyArray
```
Comment (esparta) [inline array.c:3217]: Is this correct? As long I can see `to_ary` returns an object with the same class that self:

```
ruby -e "class MyArray < Array; end; puts MyArray.new([]).to_ary.class"
MyArray
```

`to_ary` does sightly different work than `to_a` which effectively returns an `Array` instance:

```ruby
ruby -e "class MyArray < Array; end; puts MyArray.new([]).to_a.class"
Array

```

Comment (BurdetteLamar) [reply array.c:3217]: Oops.  Fixed in next push.  Thanks.

File: array.c (hunk 2 above)
Comment (BurdetteLamar) [reply array.c:3258]: Have not found a satisfactory phrasing that includes both.
Can we get by stet on the strength of "a1.equal?(a)"?

File: array.c (hunk 1 above)
Comment (BurdetteLamar) [reply array.c]: Still working this one.  Nuther PR to follow.

Comment (BurdetteLamar) [reply array.c]: I have revised the doc for #to_h.

File: array.c [hunk 4]
```diff
@@ -3097,20 +3097,22 @@ rb_ary_to_s(VALUE ary)
...
  *    a = [:foo, 'bar', 2]
+ *    a.instance_of?(Array) # => true
  *    a1 = a.to_a
- *    a1 # => [:foo, "bar", 2]
  *    a1.equal?(a) # => true # Returned self
  *
- *  If +self+ is a subclass of \Array, returns the new \Array
- *  formed by converting +self+ to an \Array:
+ *  Otherwise, when <tt>self.kind_of?(Array)</tt>,
+ *  returns a new \Array containing the actual elements (not copies) of +self+:
```
Comment (jeremyevans) [inline array.c:3107]: `Otherwise, returns an instance of \Array containing the elements of +self+`.  I don't see the value in making a distinction for actual elements vs. copies.

File: array.c [hunk 5]
```diff
@@ -3097,20 +3097,22 @@ rb_ary_to_s(VALUE ary)
  *  call-seq:
  *    to_a -> self or new_array
  *
- *  Returns +self+ if <tt>self.instance_of?(Array)</tt>:
+ *  When <tt>self.instance_of?(Array)</tt>, returns +self+:
```
Comment (jeremyevans) [inline array.c:3100]: `When self is an instance of \Array, returns +self+:`

File: array.c [hunk 6]
```diff
@@ -3126,20 +3128,80 @@ rb_ary_to_a(VALUE ary)
...
  *
- *  If a block is given, the results of the block on each element of
- *  the array will be used as pairs.
+ *  When no block is given, returns a new \Hash:
+ *    [].to_h # => {}
+ *    a = [['foo', 'zero'], ['bar', 'one'], ['baz', 'two']]
+ *    h = a.to_h
+ *    h # => {"foo"=>"zero", "bar"=>"one", "baz"=>"two"}
+ *    h.keys.first.equal?(a.first.first) # => false # Key is a copy
+ *    h.values.first.equal?(a.first.last) # => true # Value is not a copy
```
Comment (jeremyevans) [inline array.c:3154]: The whole section above should be rewritten, and sections below that discuss key copying should be rewritten.  It only applies if the first element of the 2 element array is an unfrozen string.  For frozen strings and all other objects, the key is not duplicated.  This is unrelated to the `to_h` method and should not be documented here, it's how all hashes work in regards to unfrozen string keys.

File: array.c (hunk 2 above)
Comment (jeremyevans) [reply array.c:3277]: I think `Reverses +self+ in place.` is best, but at least remove the ` (not copies)`.

File: array.c [hunk 7]
```diff
@@ -3224,12 +3290,13 @@ rb_ary_reverse_bang(VALUE ary)
...
 /*
  *  call-seq:
- *     ary.reverse    -> new_ary
- *
- *  Returns a new array containing +self+'s elements in reverse order.
+ *    array.reverse -> new_array
  *
- *     [ "a", "b", "c" ].reverse   #=> ["c", "b", "a"]
- *     [ 1 ].reverse               #=> [1]
+ *  Returns a new \Array whose elements (not copies) are in reverse order:
```
Comment (jeremyevans) [inline array.c:3295]: I think the previous wording is better and more explicit.  If you must keep the new language, at least remove the ` (not copies)` part.

File: array.c (hunk 4 above)
Comment (BurdetteLamar) [reply array.c:3107]: Thanks, @jeremyevans.  It's useful to know that elements are generally not copied.  (I have only studied the methods I've written about so far.)  I had not thought anything about copies until I ran into the copied hash key.
I'll get this cleaned up.  Again, thanks for your careful and helpful reviewing.

File: array.c (hunk 2 above)
Comment (BurdetteLamar) [reply array.c]: Fixed, per @jeremyevans.

File: array.c (hunk 5 above)
Comment (BurdetteLamar) [reply array.c]: Fixed.

File: array.c (hunk 6 above)
Comment (BurdetteLamar) [reply array.c]: Done.

File: array.c [hunk 8]
```diff
@@ -3097,20 +3097,21 @@ rb_ary_to_s(VALUE ary)
...
+ *  When +self+ is an instance of \Array, returns +self+:
  *    a = [:foo, 'bar', 2]
+ *    a.instance_of?(Array) # => true
  *    a1 = a.to_a
- *    a1 # => [:foo, "bar", 2]
  *    a1.equal?(a) # => true # Returned self
  *
- *  If +self+ is a subclass of \Array, returns the new \Array
- *  formed by converting +self+ to an \Array:
+ *  Otherwise, when <tt>self.kind_of?(Array)</tt>,
```
Comment (jeremyevans) [inline array.c:3106]: `,  when <tt>self.kind_of?(Array)</tt>`  can be removed as it is redundant.  All cases where the method can be called where self is not an instance of Array, self is a kind of Array.

File: array.c [hunk 9]
```diff
@@ -3126,20 +3127,59 @@ rb_ary_to_a(VALUE ary)
...
- *     ary.to_h {|item| block }  -> hash
+ *    array.to_h -> new_hash
+ *    array.to_h {|item| ... } -> new_hash
+ *
+ *  Returns a new \Hash formed from +self+.
+ *
+ *  In all the forms below, each 2-element \Array
+ *  is formed into a key-value pair in the new \Hash.
+ *
+ *  ====== \Array of 2-Element Arrays
```
Comment (jeremyevans) [inline array.c:3138]: This should be restructured.  The first section should discuss usage with a block, in which case the block is called and should return a 2-element array.  The second section should state if a block is not given, the array must already be composed of 2-element arrays, which are used to construct the hash.  This reduces the need to explain 4 cases (2-element array with block, 2-element array without block, other array with block, other array with block), to 3 cases (block, array of 2-element arrays without block, other array without block), which makes more sense as the behavior is the same for all cases where a block is passed.

Comment (BurdetteLamar) [reply array.c:3138]: Agree that block-first is much better.   But doesn't the block get a different  yield for 2-ele and other?

Comment (jeremyevans) [reply array.c:3138]: In your example code for both cases, you yield the element in the array (`item`).  In the array of 2-element arrays case, it's a 2-element array, and in the other case it is some other object, but the behavior is the same in both cases.

Comment (BurdetteLamar) [reply array.c:3138]: That insight came to me during lunch (BLT, pinot grigio).  Thanks.

Comment (BurdetteLamar) [reply array.c]: Fixed.

**Comment by BurdetteLamar** [comment]
Yes, @jeremyevans, accidental (careless).  I value highly all of your comments (and those of the others also).  I depend on everyone to help me avoid blunders.  As you've seen, I do make mistakes, both technical and procedural.

File: array.c (hunk 8 above)
Comment (BurdetteLamar) [reply array.c:3106]: Removed.

Comment (jeremyevans) [reply array.c:3106]: This says removed, but the changes haven't been made.  Do you need to push a commit?

Comment (BurdetteLamar) [reply array.c]: Duh, yes. When I was in the Army (back at the Dawn of Time), the instructors used to say, "Take your right ear in your right hand, and your left ear in your left hand, and pull you head out of ...."
Now pushed.
